src/Lykke.Service.BlockchainHistoryReader.AzureRepositories/Implementations/HistorySourceRepository.cs

```diff
@@ -0,0 +1,257 @@
+                );
+                
+                _chaosKitty.Meow($"{nameof(HistorySourceRepository)}-{nameof(GetAsync)}");
+
+            } while (continuationToken != null);
+
+
+            return entities
+                .Select(RestoreHistorySourceFromEntity)
+                .ToArray();
+        }
+
+        public async Task<HistorySource> GetOrCreateAsync(
+            string address,
+            string blockchainType,
+            Guid clientId)
+        {
+            var entity = await _historySources.GetOrInsertAsync
+            (
+                partitionKey: GetPartitionKey(blockchainType),
+                rowKey: GetRowKey(address),
+                createNew: () =>
+                {
+                    var historySource = HistorySource.Create
+                    (
+                        address: address,
+                        blockchainType: blockchainType,
+                        clientId: clientId
+                    );
+
+                    return ConvertHistorySourceToEntity(historySource);
+                }
+            );
+
+            return RestoreHistorySourceFromEntity(entity);
+        }
+
+        public Task UpdateAsync(
+            HistorySource historySource)
+        {
+            return _historySources.MergeAsync
+            (
+                partitionKey: GetPartitionKey(historySource.BlockchainType),
+                rowKey: GetRowKey(historySource.Address),
+                mergeAction: (entity) =>
+                {
+                    entity.HistoryUpdatedOn = historySource.HistoryUpdatedOn;
+                    entity.HistoryUpdateScheduledOn = historySource.HistoryUpdateScheduledOn;
+                    entity.LatestHash = historySource.LatestHash;
+
+                    return entity;
+                }
+            );
+        }
+
+        public async Task<HistorySource> TryGetAsync(
+            string address,
+            string blockchainType)
+        {
+            var entity = await _historySources.GetDataAsync
+            (
+                partition: GetPartitionKey(blockchainType),
+                row: GetRowKey(address)
+            );
+
+            if (entity != null)
+            {
+                return HistorySource.Restore
+                (
+                    address: entity.Address,
+                    blockchainType: entity.BlockchainType,
+                    clientId: entity.ClientId,
+                    historyUpdatedOn: entity.HistoryUpdatedOn,
+                    historyUpdateScheduledOn: entity.HistoryUpdateScheduledOn,
+                    latestHash: entity.LatestHash
+                );
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private static string GetPartitionKey(
+            string blockchainType)
+        {
+            return $"{blockchainType}-{blockchainType.CalculateHexHash32(3)}";
```

Comment:
hash of blockchain type is useless here, since blockchain type already pressented in the key. Probably you meant hash of address?